### ISSUE TITLE
[False Positive] Dataclass attribute backed by descriptor showing incorrect type in `__init__`

### ISSUE THREAD

**Issue by relsunkaev** [microsoft/pyright#2885]
**Describe the bug**
When using descriptors with dataclass fields, pyright cannot infer the correct type in generated `__init__()`.

**To Reproduce**
Using the following example (doesn't have to be `Generic`, will reproduce regardless)

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

_T = TypeVar("_T")

class Descriptor(Generic[_T]):
    value: _T

    def __get__(self, obj, objtype=None) -> _T:
        return self.value

    def __set__(self, _, value: _T):
        self.value = value

@dataclass
class DataClass:
    value: Descriptor[int] = Descriptor[int]()

DataClass(value=1)
```

VS Code will highlight `DataClass(value=1)` with message

> Argument of type "Literal[1]" cannot be assigned to parameter "value" of type "Descriptor[int]" in function "__init__"
> ‎ ‎ ‎ ‎ "Literal[1]" is incompatible with "Descriptor[int]"

**Expected behavior**
No errors

**VS Code extension or command-line**
VS Code 1.63.2
Pylance v2022.1.2-pre.1

**Comment by erictraut** [comment]
I think the current behavior is correct. No type inference is involved here. You're explicitly declaring that the type of the field `value` is `Descriptor[int]`. Its type is not `int`.

Mypy gives the same error in this case.

I think you're abusing dataclasses and descriptors here. A descriptor is generally a class variable, not an instance variable. Fields within a dataclass are instance variables.

**Comment by relsunkaev** [comment]
SQLAlchemy uses a faux `Mapped` descriptor to allow `SomeModel.field.is_(...)` and such, I tried annotating with a dataclass with it as described in their docs and that's where I got the error. I can see how that can be considered bit abusive of the dataclasses API.

**Comment by erictraut** [comment]
Perhaps something like this would work for you?

```python
@dataclass
class DataClass:
    value: ClassVar[Descriptor[int]] = Descriptor[int]()
    _value: InitVar[int]

    def __post_init__(self, _value: int):
        self.value = _value

DataClass(_value=1)
```

**Comment by relsunkaev** [comment]
Yeah I stumbled across that while tinkering with this setup but with SQLAlchemy every field is a "descriptor" which would mean writing out every field twice and at that point might as well drop the dataclass `dataclass` and write an `__init__` method.

I just thought it would be nice to support something like this since it would allow the SQLAlchemy team to use `__dataclass_transform__` in their current effort of typing their library and switching to a dataclass-like declarative API and maybe should have been filed as a "feature". I can totally see how this behavior might not be desirable though.
